pynuodb/session.py

```diff
@@ -383,17 +393,32 @@ def recv(self, doStrip=True):
             msg = msg.decode("latin-1")
         return msg
 
-    def __readFully(self, msgLength):
+    def __readFully(self, msgLength, timeout=None):
         """ Pull the entire next raw bytes message from the socket """
         msg = b''
+        old_tmout = self.__sock.gettimeout()
         while msgLength > 0:
+            if timeout is not None:
+                # It's a little wrong that this timeout applies to each recv()
+                # instead of to the entire operation; however we only use this
+                # when reading the header which will always be read in one
+                # pass anyway.
+                self.__sock.settimeout(timeout)
             try:
                 received = self.__sock.recv(msgLength)
+            except socket.timeout:
+                return None
             except IOError as e:
-                raise SessionException("Session was closed while receiving: network error %s: %s" % (str(e.errno), e.strerror if e.strerror else str(e.args)))
+                raise SessionException("Session was closed while receiving: network error %s: %s" %
+                                       (str(e.errno), e.strerror if e.strerror else str(e.args)))
+            finally:
+                if timeout is not None:
+                    self.__sock.settimeout(old_tmout)
+
             if not received:
-                raise SessionException("Session was closed while receiving msgLength=[%d] len(msg)=[%d] "
-                                       "len(received)=[%d]" % (msgLength, len(msg), len(received)))
+                raise SessionException("Session was closed while receiving msgLength=[%d] "
+                                       "len(msg)=[%d] len(received)=[%d]" %
+                                       (msgLength, len(msg), len(received)))
```

Comment:
You didn't really change this code, but I'm confused by the `msg` variable that is initialized as an empty bytes literal (`b''`). This seems like some weird detail so that the data received is encoded correctly (I suspect it isn't even necessary since we still have to do something special in Python 3, which is the only Python version that even cares about the `b` prefix, according to the [Python docs](https://docs.python.org/2/reference/lexical_analysis.html#string-literals)). But putting that detail aside, there is no reason why the user would need to see that useless value, which we include in the exception message.